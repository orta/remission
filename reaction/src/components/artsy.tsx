import * as PropTypes from "prop-types"
import * as React from "react"

/**
 * The Artsy specific props injected by the higher-order component produced by `ContextConsumer`.
 *
 * @see {@link ContextProvider}
 * @see {@link ContextConsumer}
 *
 * @todo Once this PR for `rest` https://github.com/Microsoft/TypeScript/pull/13470 lands weâ€™ll be able to not make this
 *       optional and simply remove it from the props that a component wrapped with the `ContextConsumer` HOC accepts.
 */
export interface ContextProps {
  /**
   * The currently signed-in user.
   */
  currentUser?: User
}

interface PrivateContextProps extends ContextProps {
  /**
   * This prop solely exists to verify that a consumer is nested inside a provider.
   */
  _isNestedInProvider: boolean
}

const ContextTypes: React.ValidationMap<PrivateContextProps> = {
  _isNestedInProvider: PropTypes.bool,
  currentUser: PropTypes.object,
}

/**
 * This component provides Artsy specific context props to components down the tree. Components that wish to consume
 * these props should be wrapped with the `ContextConsumer` higher-order component.
 *
 * @see {@link ContextConsumer}
 */
export class ContextProvider extends React.Component<ContextProps, null>
  implements React.ChildContextProvider<PrivateContextProps> {
  static childContextTypes = ContextTypes

  constructor(props) {
    if (React.Children.count(props.children) > 1) {
      throw new Error("A ContextProvider expects a single child.")
    }
    super(props)
  }

  getChildContext() {
    return {
      _isNestedInProvider: true,
      currentUser: this.props.currentUser,
    }
  }

  render() {
    return this.props.children as any
  }
}

/**
 * Wraps a component in a higher-order component that injects the Artsy specific context props into the given component.
 * Wrapped components are expected to be rendered in a tree that contains a `ContextProvider` component.
 *
 * The injected props are described in `ContextProps`.
 *
 * @see {@link ContextProvider}
 * @see {@link ContextProps}
 *
 * @param Component The component that needs Artsy specific context props.
 */
export function ContextConsumer<P>(
  Component: React.ComponentClass<P> | React.StatelessComponent<P>
): React.ComponentClass<P> {
  const name = Component.displayName || Component.name
  return class extends React.Component<P, void> {
    static contextTypes = ContextTypes
    static displayName = `Artsy(${name})`

    constructor(props, context: PrivateContextProps) {
      if (!context._isNestedInProvider) {
        const start = name || "A component"
        throw new Error(`${start}, which needs Artsy props, was not wrapped inside a ContextProvider component.`)
      }
      super(props, context)
    }

    render() {
      const currentUser = this.context.currentUser
      const props = Object.assign({ currentUser }, this.props)
      return <Component {...props} />
    }
  }
}
