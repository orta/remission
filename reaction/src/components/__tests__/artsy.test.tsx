import * as React from "react"
import * as renderer from "react-test-renderer"

import "jest-styled-components"

import * as Artsy from "../artsy"

const ShowCurrentUser: React.SFC<Artsy.ContextProps & { additionalProp?: string }> = props => {
  let text = props.currentUser.name
  if (props.additionalProp) {
    text = `${text} & ${props.additionalProp}`
  }
  return <div>{text}</div>
}
// This HOC adds the context to the component.
const WithCurrentUser = Artsy.ContextConsumer(ShowCurrentUser)

describe("Artsy context", () => {
  const currentUser: User = {
    id: "andy-warhol",
    accessToken: "secret",
    name: "Andy Warhol",
  }

  it("exposes the currently signed-in user", () => {
    const div = renderer
      .create(
        <Artsy.ContextProvider currentUser={currentUser}>
          <WithCurrentUser />
        </Artsy.ContextProvider>
      )
      .toJSON()
    expect(div.children[0]).toEqual("Andy Warhol")
  })

  it("passes other props on", () => {
    const div = renderer
      .create(
        <Artsy.ContextProvider currentUser={currentUser}>
          <WithCurrentUser additionalProp="friends" />
        </Artsy.ContextProvider>
      )
      .toJSON()
    expect(div.children[0]).toEqual("Andy Warhol & friends")
  })

  it("throws an error when not embedded in a context provider", () => {
    expect(() => {
      renderer.create(<WithCurrentUser />)
    }).toThrowErrorMatchingSnapshot()
  })

  it("throws an error when trying to embed more than a single child in a context provider", () => {
    expect(() => {
      renderer.create(
        <Artsy.ContextProvider>
          <WithCurrentUser />
          <div />
        </Artsy.ContextProvider>
      )
    }).toThrowErrorMatchingSnapshot()
  })

  it("has a display name for React DevTools purposes", () => {
    expect(WithCurrentUser.displayName).toEqual("Artsy(ShowCurrentUser)")
  })
})
