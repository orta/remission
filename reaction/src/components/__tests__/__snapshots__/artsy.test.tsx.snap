// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Artsy context throws an error when not embedded in a context provider 1`] = `"ShowCurrentUser, which needs Artsy props, was not wrapped inside a ContextProvider component."`;

exports[`Artsy context throws an error when trying to embed more than a single child in a context provider 1`] = `"A ContextProvider expects a single child."`;
